### ISSUE TITLE
Investigate off-by-one in first generated grammar token via MEICAR_generate_trajectories CLI

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/MEDS_EIC_AR#111]
## Summary

While landing the CLI end-to-end grammar test (PR #106), I discovered a consistent off-by-one in the first grammar token emitted by ``MEICAR_generate_trajectories``: given a prompt whose last real event is a ``SEP`` (FSM state = ``BETWEEN``, so the model should emit a program-start token next), the model emits **position 1 of the next program** instead of position 0 — e.g. ``B[1]`` instead of ``B[0]``, ``C[1]`` instead of ``C[0]``. Every token after position 0 is correct.

## Reproduction

PR #106's ``test_grammar_cli_greedy_output_is_fully_grammar_valid`` fixture reliably exhibits this. Probing the output directly:

- Pretrain a Llama/GPT-NeoX model via ``MEICAR_pretrain`` on the synthetic grammar dataset from ``tests/_grammar_meds.py``.
- Generate with ``do_sample=False`` via ``MEICAR_generate_trajectories``.
- For a subject whose prompt ends at ``[A[0..3], SEP]`` (BETWEEN state), the generated grammar tokens start with ``[B[1], B[2], SEP, A[0], A[1], A[2], A[3], SEP, ...]`` — 100% grammar-valid **if we prepend** the correct program-start ``B[0]=6`` before position 0. Explicitly verified: walking the FSM from BETWEEN through ``[B[0], B[1], B[2], SEP, A[0], ...]`` gives ``15/15 valid`` for typical held-out samples.
- The same held-out sample through the in-process test path (``Model.generate`` called directly, bypassing the CLI) does **not** exhibit this behavior under controlled prompts — the in-process tests in ``test_pattern_generation.py`` achieve strict validity.

The specific failure pattern:

```
prompt_tokens (end of prompt): [2, 3, 4, 5, 1]    # A[0..3], SEP → FSM at BETWEEN
generated tokens:              [7, 8, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, ...]
                                ^^^
                                B[1] — should be B[0]=6
```

Rest of the generated sequence is grammar-valid continuation from IN-B-at-position-1. The model has clearly learned the grammar; only the first generated position is wrong.

## Likely suspects

1. **Rolling loop first-chunk handling**: ``_rolling_generate`` may be discarding or shifting the first emitted token somewhere in its accumulator. The test triggers rolling via ``rolling_generation.max_new_tokens=30`` with ``max_seq_len=16``, so the first chunk's boundary is the likeliest location.
2. **Training vs. inference distribution mismatch**: ``MEDSTorchDataset`` might insert ``TIMELINE//START`` / other aux tokens differently at training vs. inference time, making the model see a subtly different prompt structure and producing a one-position skip. Worth verifying the exact input tensor the model receives in both paths and diffing them.
3. **``format_trajectories`` column-mapping drift**: The token-index-to-code mapping happens in ``format_trajectories.py`` via ``code_information[code_idx]``. If there's an off-by-one in how ``code_information`` is built (e.g., a vocab that skips index 0 vs. one that doesn't), it would manifest as "output is one vocab position off." Direct model output inspection via a debugging endpoint would confirm or refute this.
4. **Greedy-specific**: Worth testing whether sampling mode exhibits the same pattern. In PR #106 the greedy fixture shows the off-by-one universally; the sampling fixture's behavior should be re-examined once greedy is understood.

## Investigation steps

1. Load a grammar-pretrained checkpoint directly (not through the CLI).
2. Construct a test prompt matching what ``MEICAR_generate_trajectories`` feeds the model for a known held-out subject — same token IDs, same padding, same attention mask.
3. Call ``Model.generate(...)`` directly with ``do_sample=False`` and compare:
   - Output of the direct ``Model.generate`` call (with the same prompt the CLI assembles).
   - Output of the CLI's ``predict_step`` for the same subject.
   - The parquet rows ``format_trajectories`` writes.
4. Whichever step introduces the off-by-one is the bug.

## Impact if fixed

PR #106 has a ``_walk_generated_with_first_token_grace`` walker that grants a one-token grace at position 0 to accommodate this bug while still enforcing strict validity on the remainder. Once fixed, that walker can be simplified to a pure ``_walk_generated_strictly`` (the commented-out alternative in the same file), and the PR #106 test becomes byte-for-byte as strict as ``test_pattern_generation.py``'s in-process assertions.

Currently blocking strict greedy correctness validation of the CLI pipeline.

## Labels

`bug`, `priority:high`, `Trajectory Generation / Inference`

**Comment by mmcdermott** [comment]
## Resolved: the bug was in the fixture's `prediction_time` placement, not in the CLI pipeline

Root-cause investigation landed the fix in commit 965ab8c on PR #106. Summary:

**Not a CLI pipeline bug.** Direct `Model.generate` on a controlled prompt produces strictly grammar-correct output — I verified this by loading the pretrained checkpoint, constructing a prompt matching what the CLI feeds the model, and calling `Model.generate` both in single-chunk and rolling modes. Both return 100% grammar-valid token sequences.

**The bug was in `build_grammar_meds_dataset`'s `prediction_time` placement.** `meds-torchdata`'s `MEDSPytorchDataset._prepare_task_label_schema` uses:

```python
pl.col("time").search_sorted(pl.col("prediction_time"), side="right")
```

With `side="right"`, this **includes** events at `time == prediction_time` in the prompt, not just events strictly before. My fixture was placing `prediction_time = base_time + timedelta(hours=first_sep_idx + 1)` — exactly on the hour of the next event after the SEP. Result: that next event (e.g. `B[0]`) was silently included in the prompt. What I interpreted as "the model's first generated token is `B[1]` instead of `B[0]`" was actually "the model correctly generated `B[1]` after already seeing `B[0]` in the prompt."

**Fix**: `prediction_time = base_time + timedelta(hours=first_sep_idx, minutes=30)` — strictly between the SEP event and the next event. No event lives at that time, so `search_sorted`'s `side` argument doesn't affect the result. The prompt ends exactly at SEP, the FSM is unambiguously in `BETWEEN` state at generation-start, and the model's first generated grammar token is the correct program-start.

**Downstream consequence in PR #106**: the walker simplified back to pure `_walk_generated_strictly` (no "first token grace" workaround). Test 5 (greedy) is now byte-for-byte as strict as the in-process `test_pattern_generation.py` tests' `GrammarFSM().walk(seq) == len(seq)` bar — every generated grammar token in every held-out sample must be a valid FSM continuation from the prompt's end-state. All 9 tests pass.

**Is this a real bug in `meds-torchdata`?** The `side="right"` semantic is defensible ("include data up to and including the prediction time") but it's worth documenting that `time == prediction_time` events are context, not target — particularly since downstream users who build prompts expecting the MEDS convention of "events strictly before prediction_time" will get silently off-by-one results. I'll open a follow-up on `meds-torchdata` to either clarify the docstring or change `side` to `left`; not blocking this PR.

Closing.
